tools/mkldscript.c

```diff
@@ -1,298 +1,465 @@
-                      seg->name, seg->name, seg->name, seg->name, seg->name);
+    for (i = 0; i < seg->includesCount; i++)
+    {
+        fprintf(f,
+            "%s/%s.o: %s\n",
+            dir, seg->name, seg->includes[i].fpath
+        );
+    }
 
-        // Increment the end of the segment
-        //if (seg->fields & (1 << STMT_increment))
-            //fprintf(fout, "    . += 0x%08X;\n", seg->increment);
+    fclose(f);
+}
 
-        //fprintf(fout, "    ..%s.ovl ADDR(..%s) + SIZEOF(..%s) :\n"
-        //    /*"    ..%s.bss :\n"*/
-        //    "    {\n",
-        //    seg->name, seg->name, seg->name);
-        //fprintf(fout, "        _%sSegmentOvlStart = .;\n", seg->name);
+static void write_rom_script(const char *dir, const char *fname, const struct Segment *segments, int segment_count)
+{
+    FILE *f;
+    int i;
 
-        //for (j = 0; j < seg->includesCount; j++)
-        //    fprintf(fout, "            %s (.ovl)\n", seg->includes[j].fpath);
+    f = fopen(fname, "w");
+    if (f == NULL)
+        util_fatal_error("failed to open file '%s' for writing", fname);
+
+    /* start */
+    fprintf(f,
+        "OUTPUT_ARCH (mips)\n"
+        "\n"
+        "SECTIONS\n"
+        "{\n"
+        "\t_RomStart = 0;\n"
+        "\t_RomSize = 0;\n"
+        "\n"
+    );
+
+    for (i = 0; i < segment_count; i++)
+    {
+        const struct Segment *seg = &segments[i];
 
-        ////fprintf(fout, "        . = ALIGN(0x10);\n");
+        /* begin segment */
+        if (seg->fields & (1 << STMT_romalign))
+            fprintf(f, "\t_RomSize = ALIGN(_RomSize, %i);\n", seg->romalign);
 
-        //fprintf(fout, "        _%sSegmentOvlEnd = .;\n", seg->name);
+        fprintf(f,
+            "\t_%sSegmentRomStartTemp = _RomSize;\n"
+            "\t_%sSegmentRomStart = _%sSegmentRomStartTemp;\n",
+            seg->name,
+            seg->name, seg->name
+        );
 
-        //fprintf(fout, "\n    }\n");
+        /* placement */
+        fprintf(f, "\t..%s ", seg->name);
+        if (seg->fields & (1 << STMT_after))
+            fprintf(f, "_%sSegmentEnd ", seg->after);
+        else if (seg->fields & (1 << STMT_number))
+            fprintf(f, "0x%02X000000 ", seg->number);
+        else if (seg->fields & (1 << STMT_address))
+            fprintf(f, "0x%08X ", seg->address);
+        fprintf(f, ": AT(_%sSegmentRomStart)\n\t{\n", seg->name);
+    
+        /* content sections */
+        fprintf(f, "\t\t_%sSegmentStart = .;\n\n", seg->name);
+        fprintf(f, "\t\t%s/%s.o (.text)\n", dir, seg->name);
+        fprintf(f, "\t\t%s/%s.o (.data)\n", dir, seg->name);
+        fprintf(f, "\t\t%s/%s.o (.rodata)\n", dir, seg->name);
+
+        /* overlay relocation section */
+        fprintf(f, "\t\t_%sSegmentOvlStart = .;\n", seg->name);
+
+        if (strncmp(seg->name, "ovl_", 4) == 0)
+            fprintf(f, "\t\t%s/%s.reloc.o (.ovl)\n", dir, seg->name);
```

Comment:
I was thinking to do this but I wasn't sure if it's worth it since the OvlEnd symbol is used in calculation of the segment rom size

Reply:
So the obvious solution is to switch between
```
_%sSegmentRomSize = ABSOLUTE(_%sSegmentOvlEnd - _%sSegmentTextStart);
```
and
```
_%sSegmentRomSize = ABSOLUTE(_%sSegmentRoDataEnd - _%sSegmentTextStart);
```
as appropriate

Which looked bad to me at first, but tbh may not be so bad of an idea. mkldscript is fully in control of the script it outputs after all, including the last section linked, so I don't think this can go wrong

I actually implemented this to try it out, so I figured I'd PR it and include some other suggested changes
https://github.com/Thar0/oot/pull/7
feel free to come up with your own implementation of course